### PR TITLE
[Hub Generated] Review request for Microsoft.PlaywrightTesting to add version preview/2023-10-01-preview

### PR DESCRIPTION
**Pull request by mjmadhu**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/mjmadhu-playwrighttesting-Microsoft.AzurePlaywrightService-2023-10-01-preview...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by mjmadhu** [comment]
We're intending to move towards better folder and namespace naming convention as discussed with API Stewardship Board dropping unnecessary terms from the names like Azure and Service, hence keeping it only as "playwrighttesting".

Though we'd like to stick to the current RP Namespace "Microsoft.AzurePlaywrightService" for now since our Public preview is due this September and seems like renaming RP Namespace would need sheer updates and re-visits for sign-offs as it seems to be a reset as mentioned [here](https://stackoverflow.microsoft.com/questions/267992).

Please let us know if there's any concern with the approach, else we'll fallback to the previous convention of "azureplaywrightservice" for folder structure as well for now.

**Comment by mjmadhu** [comment]
https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16114493478
Swagger Model Validation failing due to known issue where typespec is keeping fields with visibility read as mandatory in the swagger json.
https://github.com/Azure/typespec-azure/issues/3428 created.

**Comment by mjmadhu** [comment]
Pushing for review by the time the swagger model validation issue is getting fixed for early feedback.

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp [hunk 1]
```diff
@@ -0,0 +1,175 @@
+import "@typespec/http";
+import "@typespec/rest";
+import "@typespec/versioning";
+//import "@azure-tools/typespec-providerhub";
```
Comment (mentat9) [inline specification/playwrighttesting/PlaywrightTesting.Management/main.tsp:4]: Remove

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  version: "2023-10-01-preview",
+})
+@doc("Microsoft.AzurePlaywrightService Resource Provider management API.")
+@useDependency(Azure.ResourceManager.Versions.v1_0_Preview_1)
+@useDependency(Azure.Core.Versions.v1_0_Preview_2)
+namespace Microsoft.AzurePlaywrightService;
+
+interface Operations extends Azure.ResourceManager.Operations {}
+
+#suppress "@azure-tools/typespec-azure-resource-manager/arm-resource-path-segment-invalid-chars" "Existing Template"
```
Comment (mentat9) [inline specification/playwrighttesting/PlaywrightTesting.Management/main.tsp:26]: >#suppress [](http://example.com/codeflow?start=0&length=9)

Why are these suppress statements needed?

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+  @doc("The free-trial createdAt utcDateTime.")
+  @visibility("read")
+  createdAt: utcDateTime;
+
+  @doc("The free-trial expiryAt utcDateTime.")
+  @visibility("read")
+  expiryAt: utcDateTime;
+
+  @doc("The free-trial allocated limit value.")
```
Comment (mentat9) [inline specification/playwrighttesting/PlaywrightTesting.Management/main.tsp:140]: >The free-trial allocated limit value [](http://example.com/codeflow?start=8&length=36)

Units?

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  @visibility("read")
+  allocatedValue: int32;
+
+  @doc("The free-trial used value.")
```
Comment (mentat9) [inline specification/playwrighttesting/PlaywrightTesting.Management/main.tsp:144]: >The free-trial used value [](http://example.com/codeflow?start=8&length=25)

Units?

**Comment by mentat9** [comment]
> https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16114493478 Swagger Model Validation failing due to known issue where typespec is keeping fields with visibility read as mandatory in the swagger json. [Azure/typespec-azure#3428](https://github.com/Azure/typespec-azure/issues/3428) created.

This isn't a bug. Read-only properties should be allowed in the PUT payload and their values ignored by the backend. A client should be allowed to use this pattern: GET-<modify a writeable property>-PUT, without needing to massage the payload to filter out all of the read-only properties. The GET and PUT payloads should use the same definitions.

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp (hunk 1 above)
Comment (mjmadhu) [reply specification/playwrighttesting/PlaywrightTesting.Management/main.tsp]: done

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp (hunk 2 above)
Comment (mjmadhu) [reply specification/playwrighttesting/PlaywrightTesting.Management/main.tsp]: were present in auto-generated typespec doc for provider Hub User RP. Removed them.

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp (hunk 4 above)
Comment (mjmadhu) [reply specification/playwrighttesting/PlaywrightTesting.Management/main.tsp]: this depends on the quota bucket on which free trial is applied. Added eg. for now.

File: specification/playwrighttesting/PlaywrightTesting.Management/main.tsp (hunk 3 above)
Comment (mjmadhu) [reply specification/playwrighttesting/PlaywrightTesting.Management/main.tsp]: this depends on the quota bucket on which free trial is applied. Added eg. for now.

**Comment by mjmadhu** [comment]
> > https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16114493478 Swagger Model Validation failing due to known issue where typespec is keeping fields with visibility read as mandatory in the swagger json. [Azure/typespec-azure#3428](https://github.com/Azure/typespec-azure/issues/3428) created.
> 
> This isn't a bug. Read-only properties should be allowed in the PUT payload and their values ignored by the backend. A client should be allowed to use this pattern: GET--PUT, without needing to massage the payload to filter out all of the read-only properties. The GET and PUT payloads should use the same definitions.

Shouldn't they be at least considered optional for Swagger validation in PUT requests as they're not expected in requests. Eg. https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16550083315

Let me know if we should still add a read-only property in PUT request to fix this error.

**Comment by mentat9** [comment]
> > > https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16114493478 Swagger Model Validation failing due to known issue where typespec is keeping fields with visibility read as mandatory in the swagger json. [Azure/typespec-azure#3428](https://github.com/Azure/typespec-azure/issues/3428) created.
> > 
> > 
> > This isn't a bug. Read-only properties should be allowed in the PUT payload and their values ignored by the backend. A client should be allowed to use this pattern: GET--PUT, without needing to massage the payload to filter out all of the read-only properties. The GET and PUT payloads should use the same definitions.
> 
> Shouldn't they be at least considered optional for Swagger validation in PUT requests as they're not expected in requests. Eg. https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16550083315
> 
> Let me know if we should still add a read-only property in PUT request to fix this error.

Yes, it should be optional in PUT, and you should add it to the PUT payload. You can use x-ms-mutability (https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-mutability) instead of read-only to clearly convey how this property is handled.

**Comment by mjmadhu** [comment]
> > > > https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16114493478 Swagger Model Validation failing due to known issue where typespec is keeping fields with visibility read as mandatory in the swagger json. [Azure/typespec-azure#3428](https://github.com/Azure/typespec-azure/issues/3428) created.
> > > 
> > > 
> > > This isn't a bug. Read-only properties should be allowed in the PUT payload and their values ignored by the backend. A client should be allowed to use this pattern: GET--PUT, without needing to massage the payload to filter out all of the read-only properties. The GET and PUT payloads should use the same definitions.
> > 
> > 
> > Shouldn't they be at least considered optional for Swagger validation in PUT requests as they're not expected in requests. Eg. https://github.com/Azure/azure-rest-api-specs/pull/25445/checks?check_run_id=16550083315
> > Let me know if we should still add a read-only property in PUT request to fix this error.
> 
> Yes, it should be optional in PUT, and you should add it to the PUT payload. You can use x-ms-mutability (https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-mutability) instead of read-only to clearly convey how this property is handled.

The swagger is TypeSpec generated hence created that bug. Nevertheless, I'll add the property in PUT example.

**Comment by mjmadhu** [comment]
Ironically, now I'm getting this - 
image

**Comment by mjmadhu** [comment]
MicrosoftTeams-image

Not sure if I'm missing anything for prettier check?

**Comment by mjmadhu** [comment]
azp

**Review by DominikMe — APPROVED** [review]
data-plane looks good to me (besides the open bug with the readOnly prop in the PUT request)

**Review by mikekistler — APPROVED** [review]
Looks good. 👍

**Review by DominikMe — APPROVED** [review]
Checked the new Account state property, PR still looks good to me.

**Comment by rkmanda** [comment]
Added the breaking change approval label based on the apistewardship signoff

**Comment by mentat9** [comment]
> Ironically, now I'm getting this - image

Yes, sorry about that. The check on RO properties is wrong: please go ahead and put the properties back in the PUT payload. We'll make sure your PR can merge.

**Comment by mjmadhu** [comment]
For my understanding, any reason it's required to have readonly property in PUT payload on example??

**Comment by mentat9** [comment]
> For my understanding, any reason it's required to have readonly property in PUT payload on example?? They should anyways be optional? @rkmanda @mentat9

It's not required the properties be in the PUT payload, but it's required that they be **allowed** in the payload. Hence, they are required in the swagger.

**Review by scbedd — APPROVED** [review]
The `OBJECT_MISSING_REQUIRED_PROPERTY` errors are a kafka-trap. If Madhuri fixes them, she'll immediately end up with `READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST`, which is caused by an outstanding issue with [oav.](https://github.com/Azure/azure-sdk-tools/issues/6840)

When the bug in OAV is fixed, the `READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST` with this PR will disappear. Additional work will be necessary to update the `OBJECT_MISSING_REQUIRED_PROPERTY` issues.

**Review by rkmanda — APPROVED** [review]

File: specification/playwrighttesting/PlaywrightTesting.Management/examples/Quotas_Get.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    "subscriptionId": "00000000-0000-0000-0000-000000000000",
+    "location": "eastus",
+    "name": "ScalableExecution"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "id": "/subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.AzurePlaywrightService/locations/eastus/quotas/ScalableExecution",
+        "name": "ScalableExecution",
+        "type": "Microsoft.AzurePlaywrightService/Locations/Quotas",
```
Comment (TimLovellSmith) [inline specification/playwrighttesting/PlaywrightTesting.Management/examples/Quotas_Get.json:15]: resource type names should be capitalized with camelcase
"Microsoft.AzurePlaywrightService/Locations/Quotas",

**Pull request merged**